src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockImporter.scala

```diff
@@ -0,0 +1,350 @@
+        case Running => running
+        case ResolvingMissingNode(blocksToRetry) => resolvingMissingNode(blocksToRetry)
+        case ResolvingBranch(from) => resolvingBranch(from)
+      }
+      context become behavior(newState)
+    case PickBlocks => pickBlocks(state)
+  }
+
+  private def resolvingMissingNode(blocksToRetry: NonEmptyList[Block])(state: ImporterState): Receive = {
+    case BlockFetcher.FetchedStateNode(nodeData) =>
+      val node = nodeData.values.head
+      blockchain.saveNode(kec256(node), node.toArray, Block.number(blocksToRetry.head))
+      importBlocks(blocksToRetry)(state)
+  }
+
+  private def resolvingBranch(from: BigInt)(state: ImporterState): Receive =
+    running(ImporterState.resolvingBranch(from, state))
+
+  private def start(): Unit = {
+    val currentBest = getStartingBlockNumber()
+    log.debug(
+      "Starting Regular Sync, current best block with hash is {}, best by blockchain is {}",
+      currentBest,
+      blockchain.getBestBlockNumber())
+    fetcher ! BlockFetcher.Start(self, currentBest)
+    context become running(ImporterState.initial)
+  }
+
+  private def pickBlocks(state: ImporterState): Unit = {
+    val currentBest = getStartingBlockNumber()
+    log.debug(
+      "Picking blocks, current best block with hash is {}, best by blockchain is {}",
+      currentBest,
+      blockchain.getBestBlockNumber())
+
+    val msg = state.resolvingBranchFrom.fold[BlockFetcher.FetchMsg](
+      BlockFetcher.PickBlocks(blocksBatchSize)
+    )(
+      from => BlockFetcher.StrictPickBlocks(from, currentBest)
+    )
+
+    fetcher ! msg
+  }
+
+  private def importBlocks(blocks: NonEmptyList[Block]): ImportFn =
+    importWith({
+      log.debug("Attempting to import blocks starting from {}", Block.number(blocks.head))
+      Future
+        .successful(resolveBranch(blocks))
+        .flatMap({
+          case Right(blocksToImport) => handleBlocksImport(blocksToImport)
+          case Left(resolvingFrom) => Future.successful(ResolvingBranch(resolvingFrom))
+        })
+    })
+
+  private def handleBlocksImport(blocks: List[Block]): Future[NewBehavior] =
+    tryImportBlocks(blocks)
+      .map(value => {
+        val (importedBlocks, errorOpt) = value
+        log.debug("imported blocks {}", importedBlocks.map(Block.number).mkString(","))
+
+        errorOpt match {
+          case None =>
+            self ! PickBlocks
+            Running
+          case Some(err) =>
+            log.error(
+              "block import error {}, current best block is {} and best block with hash is {}",
+              err,
+              blockchain.getBestBlockNumber(),
+              getStartingBlockNumber())
+            val notImportedBlocks = blocks.drop(importedBlocks.size)
+
+            err match {
+              case e: MissingNodeException =>
+                fetcher ! BlockFetcher.FetchStateNode(e.hash)
+                ResolvingMissingNode(NonEmptyList(notImportedBlocks.head, notImportedBlocks.tail))
```

Comment:
what if `notImportedBlocks` is empty? same in line 152

Reply:
It won't happen if there is catched error.